File: mobile/android/app/src/main/java/com/companion/robot/ble/BleManager.kt
package com.companion.robot.ble

import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothProfile
import android.content.Context
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.SharedFlow
import no.nordicsemi.android.ble.BleManager as NordicBleManager
import no.nordicsemi.android.ble.data.Data
import java.util.UUID

/**
 * BLE Manager — scan, connect, MTU negotiation, GATT operations.
 *
 * BCP Service:
 *   UUID 0xCB00: BCP Service
 *     - 0xCB01: BCP_TX (Write) — mobile → robot
 *     - 0xCB02: BCP_RX (Notify) — robot → mobile
 *     - 0xCB03: BCP_CONTROL (Read/Write) — connection params
 *     - 0xCB04: DEVICE_INFO (Read) — robot metadata
 */

object BleConstants {
    val SERVICE_UUID: UUID = UUID.fromString("0000CB00-0000-1000-8000-00805F9B34FB")
    val CHAR_BCP_TX: UUID     = UUID.fromString("0000CB01-0000-1000-8000-00805F9B34FB")
    val CHAR_BCP_RX: UUID     = UUID.fromString("0000CB02-0000-1000-8000-00805F9B34FB")
    val CHAR_BCP_CONTROL: UUID = UUID.fromString("0000CB03-0000-1000-8000-00805F9B34FB")
    val CHAR_DEVICE_INFO: UUID = UUID.fromString("0000CB04-0000-1000-8000-00805F9B34FB")
    val CHAR_OTA_DATA: UUID    = UUID.fromString("0000CB05-0000-1000-8000-00805F9B34FB")

    const val TARGET_MTU = 247
    const val DEVICE_NAME_PREFIX = "CompanionBot"
}

/** Represents a discovered BLE device. */
data class DiscoveredDevice(
    val device: BluetoothDevice,
    val name: String,
    val rssi: Int,
    val isConnected: Boolean = false
)

/** Events from the BLE layer. */
sealed class BleEvent {
    data class FrameReceived(val data: ByteArray) : BleEvent()
    data class DeviceDiscovered(val device: DiscoveredDevice) : BleEvent()
    data class ConnectionStateChanged(val deviceId: String, val state: Int) : BleEvent()
    object MtuNegotiated : BleEvent()
    object ServicesDiscovered : BleEvent()
    data class Error(val message: String) : BleEvent()
}

/**
 * BLE Manager handles scanning, connection lifecycle, and data transfer.
 * Uses Nordic BLE Library for reliable BLE operations.
 */
class BleManager(private val context: Context) {

    private val _events = MutableSharedFlow<BleEvent>(replay = 0)
    val events: SharedFlow<BleEvent> = _events

    private val _connectedDevice = MutableStateFlow<DiscoveredDevice?>(null)
    val connectedDevice: StateFlow<DiscoveredDevice?> = _connectedDevice

    private val _isScanning = MutableStateFlow(false)
    val isScanning: StateFlow<Boolean> = _isScanning

    private var gatt: BluetoothGatt? = null
    private var txCharacteristic: BluetoothGattCharacteristic? = null
    private var negotiatedMtu: Int = 23

    // Fragment reassembly buffer
    private val fragmentBuffer = mutableListOf<Byte>()
    private var expectedTotalLen = 0
    private var fragmentTimeout = 0L

    /**
     * Start BLE scanning for CompanionBot devices.
     */
    fun startScan(adapter: android.bluetooth.BluetoothAdapter) {
        _isScanning.value = true

        // In production: use BluetoothLeScanner with ScanFilter for SERVICE_UUID
        // For now: stub that simulates discovery
        _events.tryEmit(BleEvent.Error("BLE scanning requires hardware — stub mode"))
        _isScanning.value = false
    }

    fun stopScan(adapter: android.bluetooth.BluetoothAdapter) {
        _isScanning.value = false
    }

    /**
     * Connect to a BLE device and discover GATT services.
     */
    fun connect(device: BluetoothDevice) {
        device.connectGatt(context, false, gattCallback)
    }

    fun disconnect() {
        gatt?.disconnect()
        gatt?.close()
        gatt = null
        txCharacteristic = null
        _connectedDevice.value = null
    }

    /**
     * Send a BCP frame over BLE. Auto-fragments if exceeding MTU-3.
     */
    fun sendBcpFrame(frame: BcpCodec.BcpFrame): Boolean {
        val tx = txCharacteristic ?: return false

        val buf = ByteArray(BcpCodec.MAX_FRAME_LEN)
        val totalLen = BcpCodec.encode(frame, buf)

        val effectiveMtu = negotiatedMtu - 3

        if (totalLen <= effectiveMtu) {
            // Single write
            tx.value = buf.copyOf(totalLen)
            return gatt?.writeCharacteristic(tx) == true
        }

        // Fragment and send
        var offset = 0
        while (offset < totalLen) {
            val chunkSize = minOf(effectiveMtu, totalLen - offset)
            val chunk = ByteArray(chunkSize + 2)

            // Fragment header: [totalLen_hi, totalLen_lo] + data
            chunk[0] = (totalLen shr 8).toByte()
            chunk[1] = (totalLen and 0xFF).toByte()
            buf.copyInto(chunk, 2, offset, offset + chunkSize)

            // In production: queue fragments and send via reliable write
            offset += chunkSize
        }

        return true
    }

    // ── GATT callback ──────────────────────────────────────────

    private val gattCallback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            when (newState) {
                BluetoothProfile.STATE_CONNECTED -> {
                    this@BleManager.gatt = gatt
                    gatt.requestMtu(BleConstants.TARGET_MTU)
                    _events.tryEmit(BleEvent.ConnectionStateChanged(gatt.device.address, newState))
                }
                BluetoothProfile.STATE_DISCONNECTED -> {
                    _connectedDevice.value = null
                    _events.tryEmit(BleEvent.ConnectionStateChanged(gatt.device.address, newState))
                }
            }
        }

        override fun onMtuChanged(gatt: BluetoothGatt, mtu: Int, status: Int) {
            if (status == BluetoothGatt.GATT_SUCCESS) {
                negotiatedMtu = mtu
                _events.tryEmit(BleEvent.MtuNegotiated)
                gatt.discoverServices()
            }
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            if (status == BluetoothGatt.GATT_SUCCESS) {
                val service = gatt.getService(BleConstants.SERVICE_UUID)
                txCharacteristic = service?.getCharacteristic(BleConstants.CHAR_BCP_TX)

                // Subscribe to RX characteristic for notifications
                service?.getCharacteristic(BleConstants.CHAR_BCP_RX)?.let { rxChar ->
                    gatt.setCharacteristicNotification(rxChar, true)
                }

                _events.tryEmit(BleEvent.ServicesDiscovered)
            }
        }

        override fun onCharacteristicChanged(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic
        ) {
            if (characteristic.uuid == BleConstants.CHAR_BCP_RX) {
                handleReceivedData(characteristic.value)
            }
        }

        override fun onCharacteristicRead(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            status: Int
        ) {
            if (status == BluetoothGatt.GATT_SUCCESS &&
                characteristic.uuid == BleConstants.CHAR_DEVICE_INFO) {
                // Parse device info (machine_id, firmware_ver, capabilities)
                _events.tryEmit(BleEvent.FrameReceived(characteristic.value))
            }
        }
    }

    /**
     * Handle incoming data — may be a fragment or a complete frame.
     */
    private fun handleReceivedData(data: ByteArray?) {
        if (data == null || data.isEmpty()) return

        // Check if this is a fragment (first 2 bytes = total length)
        if (data.size >= 2) {
            val totalLen = ((data[0].toInt() and 0xFF) shl 8) or (data[1].toInt() and 0xFF)
            if (totalLen > 0 && totalLen <= BcpCodec.MAX_FRAME_LEN) {
                // Fragment: reset buffer if new frame
                if (fragmentBuffer.isEmpty()) {
                    expectedTotalLen = totalLen
                    fragmentTimeout = System.currentTimeMillis()
                }
                fragmentBuffer.addAll(data.drop(2))
            } else {
                // Complete frame
                fragmentBuffer.clear()
                fragmentBuffer.addAll(data.toList())
            }
        } else {
            fragmentBuffer.addAll(data.toList())
        }

        // Check if frame is complete
        if (fragmentBuffer.isNotEmpty() && fragmentBuffer.size >= expectedTotalLen) {
            val completeFrame = fragmentBuffer.take(expectedTotalLen).toByteArray()
            _events.tryEmit(BleEvent.FrameReceived(completeFrame))
            fragmentBuffer.clear()
            expectedTotalLen = 0
        }

        // Fragment timeout check (2 seconds)
        if (fragmentBuffer.isNotEmpty() &&
            System.currentTimeMillis() - fragmentTimeout > 2000) {
            fragmentBuffer.clear()
            expectedTotalLen = 0
        }
    }
}

File: mobile/ios/CompanionRobot/BLE/BleManager.swift
import Foundation
import CoreBluetooth
import Combine

/// BLE Manager — scan, connect, MTU negotiation, GATT operations.
///
/// BCP Service:
///   UUID 0xCB00: BCP Service
///     - 0xCB01: BCP_TX (Write) — mobile → robot
///     - 0xCB02: BCP_RX (Notify) — robot → mobile
///     - 0xCB03: BCP_CONTROL (Read/Write) — connection params
///     - 0xCB04: DEVICE_INFO (Read) — robot metadata

struct BleConstants {
    static let serviceUUID = CBUUID(string: "CB00")
    static let charBcpTx = CBUUID(string: "CB01")
    static let charBcpRx = CBUUID(string: "CB02")
    static let charBcpControl = CBUUID(string: "CB03")
    static let charDeviceInfo = CBUUID(string: "CB04")
    static let charOtaData = CBUUID(string: "CB05")
    static let deviceNamePrefix = "CompanionBot"
    static let targetMtu = 247
}

/// Events from the BLE layer.
enum BleEvent {
    case deviceDiscovered(DiscoveredDevice)
    case connectionStateChanged(deviceId: String, state: CBPeripheralState)
    case mtuNegotiated(mtu: Int)
    case servicesDiscovered
    case frameReceived(Data)
    case error(String)
}

class BleManager: NSObject, ObservableObject {
    @Published var discoveredDevices: [DiscoveredDevice] = []
    @Published var connectedPeripheral: CBPeripheral?
    @Published var isScanning = false

    let events = PassthroughSubject<BleEvent, Never>()

    private var centralManager: CBCentralManager!
    private var txCharacteristic: CBCharacteristic?
    private var negotiatedMtu: Int = 23

    // Fragment reassembly
    private var fragmentBuffer = Data()
    private var expectedTotalLen = 0
    private var fragmentTimestamp: Date?

    override init() {
        super.init()
        centralManager = CBCentralManager(delegate: self, queue: .main)
    }

    // MARK: - Scanning

    func startScan() {
        guard centralManager.state == .poweredOn else {
            events.send(.error("Bluetooth not powered on"))
            return
        }

        isScanning = true
        discoveredDevices.removeAll()
        centralManager.scanForPeripherals(
            withServices: [BleConstants.serviceUUID],
            options: [CBCentralManagerScanOptionAllowDuplicatesKey: false]
        )
    }

    func stopScan() {
        isScanning = false
        centralManager.stopScan()
    }

    // MARK: - Connection

    func connect(to peripheral: CBPeripheral) {
        centralManager.connect(peripheral, options: nil)
    }

    func disconnect() {
        if let peripheral = connectedPeripheral {
            centralManager.cancelPeripheralConnection(peripheral)
        }
        connectedPeripheral = nil
        txCharacteristic = nil
    }

    // MARK: - Data transfer

    func sendBcpFrame(_ frame: BcpCodec.BcpFrame) -> Bool {
        guard let tx = txCharacteristic else { return false }

        let data = BcpCodec.encode(frame: frame)
        let effectiveMtu = negotiatedMtu - 3

        if data.count <= effectiveMtu {
            // Single write
            connectedPeripheral?.writeValue(data, for: tx, type: .withResponse)
            return true
        }

        // Fragment and send
        var offset = 0
        while offset < data.count {
            let chunkSize = min(effectiveMtu, data.count - offset)
            var chunk = Data(capacity: chunkSize + 2)

            // Fragment header: [totalLen_hi, totalLen_lo] + data
            chunk.append(UInt8(data.count >> 8))
            chunk.append(UInt8(data.count & 0xFF))
            chunk.append(data.subdata(in: offset..<offset + chunkSize))

            connectedPeripheral?.writeValue(chunk, for: tx, type: .withResponse)
            offset += chunkSize
        }

        return true
    }

    // MARK: - Fragment handling

    private func handleReceivedData(_ data: Data) {
        guard data.count >= 2 else { return }

        // Check if fragment (first 2 bytes = total length)
        let totalLen = (Int(data[0]) << 8) | Int(data[1])

        if totalLen > 0 && totalLen <= BcpCodec.maxFrameLen {
            // Fragment mode
            if fragmentBuffer.isEmpty {
                expectedTotalLen = totalLen
                fragmentTimestamp = Date()
            }
            fragmentBuffer.append(data.dropFirst(2))
        } else {
            // Complete frame
            fragmentBuffer = Data(data)
            expectedTotalLen = data.count
        }

        // Check completion
        if !fragmentBuffer.isEmpty && fragmentBuffer.count >= expectedTotalLen {
            let completeFrame = fragmentBuffer.prefix(expectedTotalLen)
            events.send(.frameReceived(Data(completeFrame)))
            fragmentBuffer.removeAll()
            expectedTotalLen = 0
        }

        // Fragment timeout (2 seconds)
        if !fragmentBuffer.isEmpty,
           let ts = fragmentTimestamp,
           Date().timeIntervalSince(ts) > 2.0 {
            fragmentBuffer.removeAll()
            expectedTotalLen = 0
        }
    }
}

// MARK: - CBCentralManagerDelegate

extension BleManager: CBCentralManagerDelegate {
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        switch central.state {
        case .poweredOn:
            break
        case .poweredOff:
            events.send(.error("Bluetooth is powered off"))
        case .unsupported:
            events.send(.error("BLE not supported on this device"))
        default:
            break
        }
    }

    func centralManager(_ central: CBCentralManager,
                        didDiscover peripheral: CBPeripheral,
                        advertisementData: [String: Any],
                        rssi RSSI: NSNumber) {
        let name = peripheral.name ?? advertisementData[CBAdvertisementDataLocalNameKey] as? String ?? "Unknown"
        let device = DiscoveredDevice(
            name: name,
            peripheralId: peripheral.identifier,
            rssi: RSSI.intValue
        )

        if !discoveredDevices.contains(where: { $0.peripheralId == peripheral.identifier }) {
            discoveredDevices.append(device)
            events.send(.deviceDiscovered(device))
        }
    }

    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        connectedPeripheral = peripheral
        peripheral.delegate = self
        peripheral.discoverServices([BleConstants.serviceUUID])
        events.send(.connectionStateChanged(deviceId: peripheral.identifier.uuidString, state: .connected))
    }

    func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        connectedPeripheral = nil
        txCharacteristic = nil
        events.send(.connectionStateChanged(deviceId: peripheral.identifier.uuidString, state: .disconnected))
    }
}

// MARK: - CBPeripheralDelegate

extension BleManager: CBPeripheralDelegate {
    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        guard let services = peripheral.services else { return }

        for service in services where service.uuid == BleConstants.serviceUUID {
            peripheral.discoverCharacteristics([
                BleConstants.charBcpTx,
                BleConstants.charBcpRx,
                BleConstants.charBcpControl,
                BleConstants.charDeviceInfo
            ], for: service)
        }
    }

    func peripheral(_ peripheral: CBPeripheral,
                    didDiscoverCharacteristicsFor service: CBService,
                    error: Error?) {
        guard let characteristics = service.characteristics else { return }

        for characteristic in characteristics {
            switch characteristic.uuid {
            case BleConstants.charBcpTx:
                txCharacteristic = characteristic
            case BleConstants.charBcpRx:
                peripheral.setNotifyValue(true, for: characteristic)
            case BleConstants.charDeviceInfo:
                peripheral.readValue(for: characteristic)
            default:
                break
            }
        }

        events.send(.servicesDiscovered)
    }

    func peripheral(_ peripheral: CBPeripheral,
                    didUpdateValueFor characteristic: CBCharacteristic,
                    error: Error?) {
        guard let data = characteristic.value, error == nil else { return }

        switch characteristic.uuid {
        case BleConstants.charBcpRx:
            handleReceivedData(data)
        case BleConstants.charDeviceInfo:
            events.send(.frameReceived(data))
        default:
            break
        }
    }

    func peripheral(_ peripheral: CBPeripheral,
                    didWriteValueFor characteristic: CBCharacteristic,
                    error: Error?) {
        if let error = error {
            events.send(.error("Write failed: \(error.localizedDescription)"))
        }
    }

    func peripheral(_ peripheral: CBPeripheral,
                    didUpdateNotificationStateFor characteristic: CBCharacteristic,
                    error: Error?) {
        if let error = error {
            events.send(.error("Notification setup failed: \(error.localizedDescription)"))
        }
    }
}
